core/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -89,6 +92,34 @@ protected boolean transportCompress() {
 
     protected abstract NodesResponse newResponse(NodesRequest request, AtomicReferenceArray nodesResponses);
 
+    /**
+     * Map the responses into {@code expected} responses and {@link FailedNodeException}s.
+     *
+     * @param expected The response class expected to be found in the results
+     * @param responses The responses to map
+     * @return Never {@code null}.
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    protected <T extends BaseNodeResponse> ResponseContainer<T> mapResponses(Class<T> expected, AtomicReferenceArray responses) {
+        final List<T> nodesResponses = new ArrayList<>();
+        final List<FailedNodeException> failures = new ArrayList<>();
+
+        for (int i = 0; i < responses.length(); ++i) {
+            Object response = responses.get(i);
+
+            if (expected.isInstance(response)) {
+                nodesResponses.add(expected.cast(response));
+            } else if (response instanceof FailedNodeException) {
+                failures.add((FailedNodeException)response);
+            } else {
+                logger.warn("unexpected response [{}], expected [{}] or [{}]",
+                            response, expected.getSimpleName(), FailedNodeException.class.getSimpleName());
+            }
+        }
+
+        return new ResponseContainer<>(expected, nodesResponses, failures);
```

Comment:
I think that the `ResponseContainer` abstraction is completely unnecessary. You can just add an abstract protected method to `TransportNodesAction` that receives the collected responses and failures. Take a look at `TransportBroadcastByNodeAction` where a similar mechanism is employed (the grouping is handled in the base class, and then delegated to an abstract protected `newResponse` method to handle creating the correct response object).